lib/parser.js

```diff
@@ -148,6 +133,34 @@ function parseFromUrl(url, fetchOptions = {}, options) {
   });
 }
 
+async function dereference(parsedJSON, initialFormat, asyncapiYAMLorJSON, options) {
+  try {
+    return await refParser.dereference(options.path, parsedJSON, {
+      parse: options.parse,
+      resolve: options.resolve,
+      dereference: options.dereference,
+    });
+  } catch (err) {
+    throw new ParserError({
+      type: 'dereference-error',
+      title: err.message,
+      parsedJSON,
+      refs: findRefs(parsedJSON, err.originalPath || err.path, options.path, initialFormat, asyncapiYAMLorJSON),
+    });
+  }
+}
+
+/*
+ * In case of circular refs, this function dereferences the spec again to dereference circular dependencies
+ * Last step is to discover those and mark properly with information whey they are pointing
+*/
+async function handleCircularRefs(parsedJSON, initialFormat, asyncapiYAMLorJSON, options) {
+  options.dereference = {circular: true};
+  await dereference(parsedJSON, initialFormat, asyncapiYAMLorJSON, options);
```

Comment:
Not a big issue but `options.dereference` is modifying the input parameter. Let's try to avoid mutating parameters as much as possible:

```suggestion
  await dereference(parsedJSON, initialFormat, asyncapiYAMLorJSON, { ...options, circular: true });
```